File: samples/feature_cuda/java/FeatureCuda.java
import org.bytedeco.opencv.global.opencv_core;
import org.bytedeco.opencv.global.opencv_features2d;
import org.bytedeco.opencv.global.opencv_imgproc;
import org.bytedeco.opencv.global.opencv_videoio;
import org.bytedeco.opencv.opencv_core.*;
import org.bytedeco.opencv.opencv_cudafeatures2d.DescriptorMatcher;
import org.bytedeco.opencv.opencv_cudafeatures2d.ORB;
import org.bytedeco.opencv.opencv_videoio.VideoCapture;

public class FeatureCuda {
    public static void main(String[] args) {
        int width = 640;
        int height = 480;
        VideoCapture cap = new VideoCapture();
        cap.set(opencv_videoio.CAP_PROP_FRAME_WIDTH, width);
        cap.set(opencv_videoio.CAP_PROP_FRAME_HEIGHT, height);
        cap.open(1);

        // initialize orb and bfmatcher
        ORB orb_cuda = ORB.create(500, 1.2f, 8, 31, 0, 2, org.bytedeco.opencv.opencv_features2d.ORB.HARRIS_SCORE, 31, 20, false);
        DescriptorMatcher bfMatcher_cuda = DescriptorMatcher.createBFMatcher(opencv_core.NORM_HAMMING);
        Mat tMat = null;
        KeyPointVector tKp = null;
        GpuMat tDesc_cuda = null;
        GpuMat mask = new GpuMat();
        while (true) {
            Mat frame = new Mat();
            boolean ret = cap.read(frame);
            if (!ret) break;

            opencv_imgproc.cvtColor(frame, frame, opencv_imgproc.COLOR_BGR2GRAY);

            // detect ORB keypoint and compute description
            KeyPointVector kp = new KeyPointVector();
            GpuMat frame_cuda = new GpuMat(frame);
            GpuMat desc_cuda = new GpuMat();
            orb_cuda.detectAndCompute(frame_cuda, mask, kp, desc_cuda);
            if (tMat == null) {
                tMat = frame.clone();
                tKp = kp;
                tDesc_cuda = desc_cuda.clone();
            }

            // brute force knn match
            DMatchVectorVector matches = new DMatchVectorVector();
            bfMatcher_cuda.knnMatch(desc_cuda, tDesc_cuda, matches, 2);

            // apply Lowe's ratio test
            matches = refineMatches(matches);

            // draw matching points
            Mat matchImg = new Mat();
            opencv_features2d.drawMatchesKnn(frame, kp, tMat, tKp, matches, matchImg);

            OpenCVFX.imshow("ORB", matchImg);
            int key = OpenCVFX.waitKey();
            if (key == 27) { // ESC to exit
                break;
            } else if (key == 32) { // Space to update matching
                releaseMat(tMat);
                tMat = null;
            }

            releaseMat(frame, matchImg);
            releaseGpuMat(desc_cuda);
        }
        cap.release();
        OpenCVFX.destroyAllWindows();
    }

    private static DMatchVectorVector refineMatches(DMatchVectorVector matches) {
        DMatchVectorVector refMatches = new DMatchVectorVector();
        for (DMatchVector match : matches.get()) {
            if (match.get()[0].distance() < 0.75 * match.get()[1].distance()) {
                refMatches.push_back(match);
            }
        }
        return refMatches;
    }

    private static void releaseMat(Mat... mats) {
        for (Mat mat : mats) {
            if (mat != null) {
                mat.release();
                mat.close();
            }
        }
    }

    private static void releaseGpuMat(GpuMat... mats) {
        for (GpuMat mat : mats) {
            if (mat != null && !mat.isNull()) {
                mat.release();
                mat.close();
            }
        }
    }
}

File: samples/feature_cuda/cpp/src/feature_cuda.cpp
#include <iostream>
#include <opencv2/opencv.hpp>
#include <opencv2/cudafeatures2d.hpp>

std::vector<std::vector<cv::DMatch>> refineMatches(std::vector<std::vector<cv::DMatch>> matches);

int main(int argc, char *argv[])
{
    int width = 640;
    int height = 480;
    cv::VideoCapture cap;
    cap.set(cv::CAP_PROP_FRAME_WIDTH, width);
    cap.set(cv::CAP_PROP_FRAME_HEIGHT, height);
    cap.open(0);

    // initailize orb and bfmatcher
    cv::Ptr<cv::FeatureDetector> orb_cuda = cv::cuda::ORB::create(500, 1.2, 8, 31, 0, 2, cv::ORB::HARRIS_SCORE, 31, 20, false);
    cv::Ptr<cv::cuda::DescriptorMatcher> bfMatcher_cuda = cv::cuda::DescriptorMatcher::createBFMatcher(cv::NORM_HAMMING);
    cv::Mat tMat;
    std::vector<cv::KeyPoint> tKp;
    cv::cuda::GpuMat tDesc_cuda;
    while (true)
    {
        cv::Mat frame;
        cap >> frame;
        if (frame.empty())
        {
            break;
        }
        cv::cvtColor(frame, frame, cv::COLOR_BGR2GRAY);

        // detect ORB keypoint and compute description
        std::vector<cv::KeyPoint> kp;
        std::vector<std::vector<cv::DMatch>> matches;
        cv::cuda::GpuMat desc;
        cv::cuda::GpuMat frame_cuda(frame);
        orb_cuda->detectAndCompute(frame_cuda, cv::cuda::GpuMat(), kp, desc);
        if (tMat.empty())
        {
            tMat = frame.clone();
            tKp = kp;
            tDesc_cuda = desc.clone();
        }

        // brute force knn match
        bfMatcher_cuda->knnMatch(desc, tDesc_cuda, matches, 2);

        // apply Lowe's ratio test
        matches = refineMatches(matches);

        // draw matching points
        cv::Mat matchImg;
        cv::drawMatches(frame, kp, tMat, tKp, matches, matchImg);

        cv::imshow("ORB", matchImg);
        int key = cv::waitKey(10);
        if (key == 27)
        { // ESC to exit
            break;
        }
        else if (key == 32)
        { // Space to update matching
            tMat.release();
        }
    }
    cap.release();
    cv::destroyAllWindows();
    return 0;
}

std::vector<std::vector<cv::DMatch>> refineMatches(std::vector<std::vector<cv::DMatch>> matches)
{
    std::vector<std::vector<cv::DMatch>> refMatches;
    for (auto match : matches)
    {
        if (match[0].distance < 0.75 * match[1].distance)
        {
            refMatches.push_back(match);
        }
    }
    return refMatches;
}
